### PR TITLE
[DNM] feat(365) - 3: add functions and docs for implementing multiple scm

### PR DESCRIPTION
**Pull request by tk3fftk**
This PR adds `getScmContext` and `canHandleUrl` function to scm-base for the multiple scm feature. 
Also, I added these documents and `_getPrInfo` document (it did not exist).

Related: https://github.com/screwdriver-cd/screwdriver/issues/365#issuecomment-313277401

Before merging for test success: https://github.com/screwdriver-cd/data-schema/pull/155

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -302,6 +302,48 @@ The list of objects consist of PR names and refs (either a branch or a sha) for
 1. Resolve with the list of objects consists of PR names and refs
 2. Reject if the input or output is not valid
 
+### getPrInfo
```
Comment (tkyi) [inline README.md]: 👍 😄

File: index.js [hunk 2]
```diff
@@ -344,6 +344,35 @@ class ScmBase {
...
+    _getScmContext() {
+        return Promise.reject('Not implemented');
+    }
+
+    /**
+     * Determine a scm module can handle a specified scm url
+     * @method canHandleUrl
+     * @return {Promise}
+     */
+    canHandleUrl(config) {
```
Comment (tkyi) [inline index.js]: Can you add JSdocs for the `config` object here as well?

**Comment by s-yoshika** [comment]
Sorry, I’ll update the issue document, it miss some points.
* `getScmContext` requires `scmUri` for it’s argument because `scm-router` should assign the scm module appropriately. To select the correct scm module, `scm-router` calls `canHandleUrl` for each scm modules and if `canHandlUrl` returns `true` in a scm module, then `scm-router` calls `getScmContext` for the scm module.
* `getDisplayName(scmContext)` is required for `scm-base`. Because, Screwdriver use `scmContext` internally, but `scmContext` is too long to display the name for users, so Screwdriver have to lookup the `displayName` ( which is `scms` setting) by `scmContext`.

File: index.js (hunk 2 above)
Comment (tk3fftk) [inline index.js]: Added!

**Comment by tk3fftk** [comment]
Screwdriver should fail until merging https://github.com/screwdriver-cd/data-schema/pull/155

File: index.js [hunk 3]
```diff
@@ -344,6 +344,45 @@ class ScmBase {
     stats() {
         return {};
     }
+
+    /**
+     * Get an array of scm context (e.g. [github.com, mygitlab_gitlab])
+     * @method getScmContext
+     * @return {Promise}
+     */
+    getScmContext() {
```
Comment (kumada626) [inline index.js]: The function name is changed from `getScmContext` to `getScmContexts`. 
https://github.com/screwdriver-cd/screwdriver/issues/365#issuecomment-314705743

Comment (tk3fftk) [inline index.js]: fixed

File: README.md [hunk 4]
```diff
@@ -302,6 +302,56 @@ The list of objects consist of PR names and refs (either a branch or a sha) for
...
+
+#### Expected Outcome
+The received webhook is available or not as boolean.
+
+#### Expected Promise Response
+1. Resolve with the received webhook is available or not.
+2. Reject if the input or output is not valid
+
+### getDisplayName
+No parameters are required.
```
Comment (wahapo) [inline README.md]: scm-router need `config.scmContext`.

Comment (tk3fftk) [inline README.md]: scm-router should need `scmContext` but other scm modules should not need. I think scm-router is a exception.

Comment (wahapo) [inline README.md]: Exceptions make it impossible to replace with other modules. I think that it is necessary to share the interface, is there any effect on this?

File: index.js [hunk 5]
```diff
@@ -344,6 +344,43 @@ class ScmBase {
...
         return {};
     }
+
+    /**
+     * Get an array of scm context (e.g. [github.com, mygitlab_gitlab])
+     * @method getScmContexts
+     * @return {Array}
+     */
+    getScmContexts() {
+        return this._getScmContexts();
```
Comment (tkyi) [inline index.js]: Should we validate that this is an array?

File: README.md [hunk 6]
```diff
@@ -302,6 +302,60 @@ The list of objects consist of PR names and refs (either a branch or a sha) for
...
+| payload        | Object | The webhook payload received from the SCM service |
+
+#### Expected Outcome
+The received webhook is available or not as boolean.
+
+#### Expected Promise Response
+1. Resolve with the received webhook is available or not.
+2. Reject if the input or output is not valid
+
+### getDisplayName (overriding needs only the case of `scm-router`)
```
Comment (tkyi) [inline README.md]: What does this mean? `(overriding needs only the case of `scm-router`)`

Is scmContext only used for `scm-router`? Could it be an optional parameter?

**Comment by catto** [comment]
need to pass scmContext to getCheckoutCommand
https://github.com/screwdriver-cd/scm-base/pull/43/files#diff-168726dbe96b3ce427e7fedce31bb0bcR132

File: README.md (hunk 6 above)
Comment (tk3fftk) [inline README.md]: `scm-github` and other scm modules use getDisplayName() as function of `scm-base` . `scm-router` only overrides this function

> Is scmContext only used for scm-router? Could it be an optional parameter?
 
Yes. We do not need scmContext and scmContext is ignored if it's added when we use single scm module. 
But we want to call the same interface which is `getDisplayName(scmContext)` in code of screwdriver API so we added the comment.

Multiple module case:
UI -> API -> getDisplayName(scmContext) of `scm-router` -> getDisplayName()  of `scm-base` of a specified module.

Single module case:
UI -> API -> getDisplayName(scmContext) of a `scm-module` -> getDisplayName()  of `scm-base` of the `scm-module`.

File: index.js (hunk 5 above)
Comment (tk3fftk) [inline index.js]: validate added!

**Comment by tk3fftk** [comment]
Added scmContext to getCheckoutCommand.

**Comment by yokawara** [comment]
Need to add scmContext to test configs.

**Comment by tk3fftk** [comment]
@yokawara added.

File: index.js [hunk 7]
```diff
@@ -344,6 +357,45 @@ class ScmBase {
...
         return {};
     }
+
+    /**
+     * Get an array of scm context (e.g. [github.com, mygitlab_gitlab])
+     * @method getScmContexts
+     * @return {Array}
+     */
+    getScmContexts() {
+        return validate(this._getScmContexts(), Joi.array().items(
```
Comment (yokawara) [inline index.js]: `getScmContexts()` should return array value, so can not use this validate method here.
This vlidate method returns Promise object.

**Comment by yokawara** [comment]
also need to addd `scmContext` to belows 🙏 
https://github.com/screwdriver-cd/scm-base/blob/master/test/index.test.js#L523
https://github.com/screwdriver-cd/scm-base/blob/master/test/index.test.js#L602

File: index.js (hunk 7 above)
Comment (tk3fftk) [inline index.js]: Thanks, I added no Promise validation.

File: index.js [hunk 8]
```diff
@@ -7,18 +7,19 @@ const dataSchema = require('screwdriver-data-schema');
...
- * @param  {Object}    config       Configuration
- * @param  {Object}    schema       Joi object used for validation
- * @return {Promise}
+ * @param  {Object}    config           Configuration
+ * @param  {Object}    schema           Joi object used for validation
+ * @param  {Boolean}   notPromise       If true, return Object
+ * @return {Promise|Object}
  */
-function validate(config, schema) {
+function validate(config, schema, notPromise) {
```
Comment (petey) [inline index.js]: Please do not have dangling booleans in a method signature. Why is this needed?

File: index.js [hunk 9]
```diff
@@ -344,6 +358,45 @@ class ScmBase {
...
+
+    /**
+     * Get an array of scm context (e.g. [github.com, mygitlab_gitlab])
+     * @method getScmContexts
+     * @return {Array}
+     */
+    getScmContexts() {
+        return validate(this._getScmContexts(), Joi.array().items(
+                Joi.reach(dataSchema.models.pipeline.base, 'scmContext').required()
+            ), true);
```
Comment (petey) [inline index.js]: Please refactor to use promises.

File: index.js [hunk 10]
```diff
@@ -315,14 +328,15 @@ class ScmBase {
     /**
      * Resolve a pull request object based on the config
      * @method getPrRef
```
Comment (tkyi) [inline index.js]: getPrInfo*

Comment (tk3fftk) [inline index.js]: fixed!

File: index.js (hunk 9 above)
Comment (tk3fftk) [inline index.js]: fixed and removed dangling boolean of validate().

File: index.js [hunk 11]
```diff
@@ -14,11 +14,7 @@ const dataSchema = require('screwdriver-data-schema');
 function validate(config, schema) {
     const result = Joi.validate(config, schema);
 
-    if (result.error) {
-        return Promise.reject(result.error);
-    }
-
-    return Promise.resolve(config);
+    return result.error ? Promise.reject(result.error) : Promise.resolve(config);
```
Comment (tk3fftk) [inline index.js]: refactor

**Comment by tk3fftk** [comment]
The scm-router uses `getScmContexts` in its constructor, so it is hard to make sure all modules are loaded before return of the constructor the case `getScmContexts` returns Promise. So we call Joi.validate() directly.

File: index.js [hunk 12]
```diff
@@ -51,6 +47,7 @@ class ScmBase {
      * @param  {String}     config.scmUri      SCM URI to add the webhook to
      * @param  {String}     config.token       Service token to authenticate with the SCM service
      * @param  {String}     config.webhookUrl  The URL to use for the webhook notifications
+     * @param  {String}     config.scmContext  The scm context name
```
Comment (minz1027) [inline index.js]: `config.scmContext` is optional right? then it should be `[config.scmContext]`.

File: index.js [hunk 13]
```diff
@@ -344,6 +353,48 @@ class ScmBase {
...
+    getScmContexts() {
+        const result = this._getScmContexts();
+        const validateResult = Joi.validate(result, Joi.array().items(
+            Joi.reach(dataSchema.models.pipeline.base, 'scmContext').required()
+        ));
+
+        return validateResult.error || result;
+    }
+
+    _getScmContexts() {
```
Comment (ian-fox) [inline index.js]: Does this get implemented in the SCM-specific repos then?

File: index.js (hunk 12 above)
Comment (tk3fftk) [inline index.js]: thanks! fixed!
